prusa-slicer: 2.8.0 -> 2.9.0

me-and opened the pull request:
Update Prusa Slicer to the latest v2.9.0. This builds on @cbat98's as-yet unmerged PR for v2.8.1 in #358215.

Changelog:

- https://files.prusa3d.com/?latest=slicer-stable&lng=en 
- https://github.com/prusa3d/PrusaSlicer/releases/tag/version_2.9.0

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

cbat98 committed:
prusa-slicer: Update version and hash

cbat98 committed:
prusa-slicer: Set opencascade version

cbat98 committed:
prusa-slicer: Update dependency with version

cbat98 committed:
prusa-slicer: Remove existing patches

cbat98 committed:
Add override for occt 7.6.1

cbat98 committed:
prusa-slicer: Use new overriden occt vn

cbat98 committed:
prusa-slicer: Move boost log link from patch to args

me-and committed:
prusa-slicer: 2.8.1 -> 2.9.0

At the same time fix the paths in a manual patch, and fix the version of
Boost to meet the build requirements.

me-and committed:
super-slicer: handle missing postPatch

When overriding the postPatch attribute from prusa-slicer, handle the
scenario where prusa-slicer doesn't have a postPatch attribute, as well
as the scenario where it does.

cbat98 commented:
Thank you for taking a look at the 2.9.0 update :) Appreciate you building on my attempt at 2.8.1.

jmickelin commented:
Nice stuff! I tested it and it works :+1: 

@maxammann raised a point in the previous PR that PrusaSlicer exhibits a bug with downloading for Curl versions newer than 7.88.1. Seeing how it's already broken in the current `2.8.0` package in `nixpkgs`, I think we can leave it as-is for this PR and then discuss overriding the Curl dependency in a followup PR.

me-and commented:
I am getting a breakage in the super-slicer-beta package, which I've not yet debugged further, or even checked if it exists prior to this change...

```
CMake Error at src/occt_wrapper/CMakeLists.txt:22 (find_package):
  Could not find a configuration file for package "OpenCASCADE" that is
  compatible with requested version "7.6.2".

  The following configuration files were considered but not accepted:

    /nix/store/yqk4xxvw5pmjjq3ah4dahka5i8yxw0z6-opencascade-occt-7.6.1/lib/cmake/opencascade/OpenCASCADEConfig.cmake, version: 7.6.1



-- Configuring incomplete, errors occurred!
```

jmickelin commented:
The issue comes from the following line: https://github.com/supermerill/SuperSlicer/blob/master/src/occt_wrapper/CMakeLists.txt#L22

It's really strange that the `super-slicer` derivation is coupled to the `prusa-slicer` one like this, since the former is lagging behind a lot still. They really should be independent packages with their own build requirements these days.

In any case, since `super-slicer` hardcodes the dependency on OCCT-7.6.2 and `prusa-slicer` pins it to 7.6.1, I think you need to override it "back" in `super-slicer.nix`, something like this:

```diff
index 66be3af..e752b8e 100644
--- a/super-slicer.nix
+++ b/super-slicer.nix  
@@ -6,6 +6,7 @@
   wxGTK31,
   prusa-slicer,
   libspnav,
+  opencascade-occt_7_6
 }:
 let
   appname = "SuperSlicer";
@@ -129,7 +130,10 @@ let
       fetchSubmodules = true;
     };
   });
-  prusa-slicer-wxGTK-override = prusa-slicer.override { wxGTK-override = wxGTK31-prusa; };
+  prusa-slicer-wxGTK-override = prusa-slicer.override { 
+    wxGTK-override = wxGTK31-prusa; 
+    opencascade-occt_7_6_1 = opencascade-occt_7_6;
+  };
   allVersions = builtins.mapAttrs (
     _name: version: (prusa-slicer-wxGTK-override.overrideAttrs (override version))
   ) versions;
```

It looks a bit hacky, with the argument `opencascade-occt_7_6_1` being set to (effectively) `opencascade-occt-7_6_2` but I think it's inevitable when it's coupled like this. The other option would be to patch SuperSlicer to downgrade the required version. Surely it would encounter the same bug that caused PrusaSlicer to pin it in the first place, so it could make sense: https://github.com/prusa3d/PrusaSlicer/commit/c6a02106fd1d3caa9a48a6b7c2bdd04546b24485#diff-54d1bb7ea7d9b83fb3c587bc042046732a50bb843a067cd86c445bafbb6ba1eeR7

cbat98 commented:
I don't know if it's something wacky with my setup, but to get it to build _for me_, I had to also apply this patch:

diff --git a/src/slic3r-arrange/include/arrange/DataStoreTraits.hpp b/src/slic3r-arrange/include/arrange/DataStoreTraits.hpp
index ba932d57a..624569d3f 100644
--- a/src/slic3r-arrange/include/arrange/DataStoreTraits.hpp
+++ b/src/slic3r-arrange/include/arrange/DataStoreTraits.hpp
@@ -71,7 +71,7 @@ template<class T, class TT = T> using WritableDataStoreOnly = std::enable_if_t<I
 template<class T, class ArrItem>
 void set_data(ArrItem &itm, const std::string &key, T &&data)
 {
-    WritableDataStoreTraits<ArrItem>::template set(itm, key, std::forward<T>(data));
+    WritableDataStoreTraits<ArrItem>::template set<T>(itm, key, std::forward<T>(data));
 }
 
 template<class T> constexpr bool IsReadWritableDataStore = IsDataStore<T> && IsWritableDataStore<T>;

I'm not sure why this was needed, but I was seeing an error in the build at about 55%.

me-and commented:
I tried to split out the super-slicer package so it no longer had the dependency on prusa-slicer, but I couldn't work out how to get that working. I _think_ the complicating factor is the use of `darwin.apple_sdk_11_0.callPackage` in `all-packages.nix`, but in honesty I didn't care enough to spend a long time on it.

In any case, @jmickelin's approach worked for me, although I modified it slightly to emulate the existing override of wxGTK in super-slicer. I'll push that patch shortly.

@cbat98 I'm baffled that you're seeing that error when it's working fine for me building out of the nixpkgs tree. Possibly you have an overlay confusing things? How do you get on with `nix-shell -p nixpkgs-review --run 'nixpkgs-review pr 367376'`, which I think enforces a clean environment?

me-and committed:
super-slicer: fix build broken by prusa-slicer update

super-slicer is built by overriding prusa-slicer, and the updated
libraries as part of the update of prusa-slicer to v2.9.0 broke
super-slicer's pinned libraries.

Ideally we'd split off super-slicer as a separate package rather than
continuing to make changes in prusa-slicer then undo them in
super-slicer, but that seems to be a much more substantial job, so stick
with the simple approach for now.

me-and commented:
...oh, that's odd. I'm now seeing a build failure as well, which looks like the same build failure as @cbat98 saw. I was building at the tip of my branch, but it looks like the build is broken by something else that's changed since I branched off...

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>prusa-slicer</li>
    <li>prusa-slicer.debug</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 5 packages built:</summary>
  <ul>
    <li>opencascade-occt_7_6_1</li>
    <li>super-slicer (super-slicer-latest)</li>
    <li>super-slicer-beta</li>
    <li>super-slicer-beta.debug</li>
    <li>super-slicer.debug (super-slicer-latest.debug)</li>
  </ul>
</details>

me-and commented:
...that was daft. I backed out 910f66b28b86f4747973d69f3c7693cdff3c7c26 because I'd interpreted the error I saw to mean the patch wasn't necessary. Turns out it just means the file the patch was applying to was in a different file.

Assuming it doesn't find any other problems, I'll push an updated branch as soon as my local build has finished.

me-and commented:
There we go!

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 7 packages built:</summary>
  <ul>
    <li>opencascade-occt_7_6_1</li>
    <li>prusa-slicer</li>
    <li>prusa-slicer.debug</li>
    <li>super-slicer (super-slicer-latest)</li>
    <li>super-slicer-beta</li>
    <li>super-slicer-beta.debug</li>
    <li>super-slicer.debug (super-slicer-latest.debug)</li>
  </ul>
</details>

cbat98 reviewed (commented):
Looks good - builds and runs perfectly on my system :D

mkg20001 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 367376`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 7 packages built:</summary>
  <ul>
    <li>opencascade-occt_7_6_1</li>
    <li>prusa-slicer</li>
    <li>prusa-slicer.debug</li>
    <li>super-slicer (super-slicer-latest)</li>
    <li>super-slicer-beta</li>
    <li>super-slicer-beta.debug</li>
    <li>super-slicer.debug (super-slicer-latest.debug)</li>
  </ul>
</details>

TilCreator commented:
Could this be backported once  #374850 is fixed?

me-and commented:
> Could this be backported once #374850 is fixed?

I don't think so: there are some breaking changes in this release, and while they're minor, I think they mean it doesn't meet the [criteria for changes acceptable for a release](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#changes-acceptable-for-releases).

If you want to stay on NixOS 24.11 but use PrusaSlicer v2.9.0, I expect you'll need to use an overlay. Two options:

* Just back-apply the changes in this PR to your current nixpkgs. I have an overlay that does more-or-less this at https://github.com/me-and/nixcfg/blob/c04e6fd7a097f5c43266fb202c5e5abd3723d538/overlays/prusa.nix. That will almost certainly mean you need to build PrusaSlicer locally, which should work but will take a bunch of CPU cycles.
* Pick up this package with its current dependencies from an unstable branch. I have an overlay that does that at https://github.com/me-and/nixcfg/blob/4e65ce2239a37c218f5623c7b740811e8be2700e/overlays/prusa.nix (although that also depends on a different and decidedly unpleasant from a Nix purity perspective overlay at https://github.com/me-and/nixcfg/blob/4e65ce2239a37c218f5623c7b740811e8be2700e/overlays/stability.nix). That should let you download a pre-built derivation rather than needing to build locally, but you'll probably need to download a lot more because you'll also need all the dependencies from the unstable branch rather than using the ones you already have in your local Nix store.
